transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java

```diff
@@ -0,0 +1,161 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+import io.netty.util.IllegalReferenceCountException;
+
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * Base class for {@link FileRegion} implementations which based on a {@link FileChannel}.
+ */
+public abstract class FileChannelBasedFileRegion implements SliceableFileRegion {
+
+    private long transferIndex;
+
+    @Override
+    public long transferBytesTo(WritableByteChannel target, long length) throws IOException {
+        long written = transferBytesTo(target, transferIndex, length);
+        if (written > 0) {
+            transferIndex += written;
+        }
+        return written;
+    }
+
+    @Override
+    public long transferIndex() {
+        return transferIndex;
+    }
+
+    @Override
+    public FileChannelBasedFileRegion transferIndex(long index) {
+        if (index < 0 || index > count()) {
+            throw new IndexOutOfBoundsException(
+                    String.format("transferIndex: %d (expected: 0 <= transferIndex <= count(%d))",
+                            index, count()));
+        }
+        this.transferIndex = index;
+        return this;
+    }
+
+    @Override
+    public boolean isTransferable() {
+        return transferIndex < count();
+    }
+
+    @Override
+    public long transferableBytes() {
+        return count() - transferIndex;
+    }
+
+    @Override
+    public long transferBytesTo(WritableByteChannel target, long position, long length)
+            throws IOException {
+        if (position < 0 || position > count()) {
+            throw new IllegalArgumentException("position out of range: " + position
+                    + " (expected: 0 - " + (this.count() - 1) + ')');
+        }
+        if (length < 0) {
+            throw new IllegalArgumentException("negative length " + length);
+        }
+        length = Math.min(count() - position, length);
+        if (length == 0) {
+            return 0;
+        }
+        open();
+        return channel().transferTo(position() + position, length, target);
+    }
+
+    @Override
+    public SliceableFileRegion slice(long index, long length) {
+        return new FileChannelBasedSlicedFileRegion(this, index, length);
+    }
+
+    @Override
+    public SliceableFileRegion retainedSlice(long index, long length) {
+        SliceableFileRegion sliced = slice(index, length);
```

Comment:
typically we override the methods which return the same type (e.g. `retain`, `touch`, etc...) so we can do things such as `return slice(index, length).retain();`. Is there any reason why we are not doing that here?

Reply:
The `retain` and `touch` method in `DefaultFileRegion` return `FileRegion` and this is class is extended by `DefaultFileRegion` so we can not declared a `retain` or `touch` method which returns a `SliceableFileRegion`. And as @normanmaurer said above, users may extend `DefaultFileRegion` so we can not change the return value type of these methods in `DefaultFileRegion`. Maybe we could do this when releasing 4.2?

Reply:
thanks for clarifying ... lgtm